x-pack/plugins/apm/public/components/shared/TransactionActionMenu/TransactionActionMenu.tsx

```diff
@@ -0,0 +1,194 @@
+  return (
+    <EuiButtonEmpty iconType="arrowDown" iconSide="right" onClick={onClick}>
+      Actions
+    </EuiButtonEmpty>
+  );
+}
+
+interface Props {
+  readonly transaction: Transaction;
+  readonly location: Location;
+}
+
+interface State {
+  readonly isOpen: boolean;
+}
+
+export class TransactionActionMenu extends React.Component<Props, State> {
+  public state: State = {
+    isOpen: false
+  };
+
+  public toggle = () => {
+    this.setState(state => ({ isOpen: !state.isOpen }));
+  };
+
+  public close = () => {
+    this.setState({ isOpen: false });
+  };
+
+  public getInfraActions(transaction: Transaction) {
+    const hostName = get(transaction, 'context.system.hostname');
+    const podId = get(transaction, 'kubernetes.pod.uid');
+    const containerId = get(transaction, 'container.id');
+    const pathname = '/app/infra';
+    const time = new Date(transaction['@timestamp']).getTime();
+    const infraMetricsQuery = getInfraMetricsQuery(transaction);
+
+    return [
+      {
+        icon: 'loggingApp',
+        label: 'Show pod logs',
+        target: podId,
+        hash: `/link-to/pod-logs/${podId}`,
+        query: { time }
+      },
+
+      {
+        icon: 'loggingApp',
+        label: 'Show container logs',
+        target: containerId,
+        hash: `/link-to/container-logs/${containerId}`,
+        query: { time }
+      },
+
+      {
+        icon: 'loggingApp',
+        label: 'Show host logs',
+        target: hostName,
+        hash: `/link-to/host-logs/${hostName}`,
+        query: { time }
+      },
+
+      {
+        icon: 'infraApp',
+        label: 'Show pod metrics',
+        target: podId,
+        hash: `/link-to/pod-detail/${podId}`,
+        query: infraMetricsQuery
+      },
+
+      {
+        icon: 'infraApp',
+        label: 'Show container metrics',
+        target: containerId,
+        hash: `/link-to/container-detail/${containerId}`,
+        query: infraMetricsQuery
+      },
+
+      {
+        icon: 'infraApp',
+        label: 'Show host metrics',
+        target: hostName,
+        hash: `/link-to/host-detail/${hostName}`,
+        query: infraMetricsQuery
+      }
+    ]
+      .filter(({ target }) => Boolean(target))
+      .map(({ icon, label, hash, query }, index) => {
+        const href = getKibanaHref({
+          location,
+          pathname,
+          hash,
+          query
+        });
+
+        return (
+          <EuiContextMenuItem icon={icon} href={href} key={index}>
+            <EuiFlexGroup gutterSize="s">
+              <EuiFlexItem>
+                <EuiLink>{label}</EuiLink>
+              </EuiFlexItem>
+              <EuiFlexItem grow={false}>
+                <EuiIcon type="popout" />
+              </EuiFlexItem>
+            </EuiFlexGroup>
+          </EuiContextMenuItem>
+        );
+      });
+  }
+
+  public render() {
+    const { transaction, location } = this.props;
+    return (
+      <QueryWithIndexPattern query={getDiscoverQuery(transaction)}>
```

Comment:
It's pretty crazy that doing async rendering has always been this hard. Hooks cannot come soon enough...

Reply:
Btw. good job. This is definitely a cleaner solution!